### DISCUSSION TITLE
Welcome to PyMuPDF Discussions!

### DISCUSSION THREAD

**Discussion by JorjMcKie** [pymupdf/PyMuPDF#752] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by JorjMcKie** [comment]
**_Jorj McKie_** is just a pseudonym - chosen after one of the characters in "The Dosadi Experiment" by Frank Herbert, one of my favorite science fiction authors. For many of you, he may be better known as the author of "Dune", one of the best SF novels ever written.
My real name is Harald Lieder, I am German and since my retirement I am living in beautiful Venezuela, Isla Margarita.
I studied Mathematics and Physics at the Franfkurt university and received a masters degree in Mathematics ("Diploma") in 1976.
My professional background is IT. Most of the time I worked for banks and insurance companies, but my most important employer ever certainly was Accenture, one of the world's largest and most important consulting companies, for whom I acquired and realized large IT infrastructure programmes like data center consolidation in post-merger situations, re-platforming, availability improvements, technical architecture and quality assurance.

As an avid reader of science publications, I was molested by the missing table of contents of one of my favorite magazines (PDF fomat) and started looking around for solutions to add these programmatically. Around 2015 / 2016 I came across fitz-python and joined maintaining it. Working with the creator @rk700 we renamed the repo to PyMuPDF, published it on PyPI and finally turned it to a GitHub "organization" as you see it today.

**Comment by FernandoJCabral** [comment]
Jorj, I need your help, please. I'm using PyMuPDF and PySimpleGUI to display a PDF page as image with a moving rectangle. In PyMuPDF this is the statements of interest:
```
def get_page(doc, page_count, dlist_tab, pno=1, x1=250, y1=600):
    dlist = dlist_tab[pno]  # get display list
    pagina = doc[pno]
    r1 = fitz.Rect(x1, y1, x1+160, y1+100)
    shape = pagina.new_shape()
    shape.draw_rect(r1)  # draw rectangles
    gold = (1,1,0)
    shape.finish(width = 2.5, color=gold)
    shape.commit()
    dlist=pagina.getDisplayList()
    r = dlist.rect  # page rectangle
    zoom_x = 0.7  # horizontal zoom
    zoom_y = 0.7  # vertical zoom
    mat = fitz.Matrix(zoom_x, zoom_y)  # zoom factor 0.7 in each dimension
    pix = dlist.getPixmap(alpha=False, matrix=mat)
    return pix.getPNGData()  # return the PNG image
```
So, basicaly I get a PDF page, draw a rectangle on top of it and create a PNG image.
That PNG image is displayed using PySimpleGUI with this statement:
```    image_elem = sg.Image(data=data)```
(data is the PNG image).
Now the user can make the rectangle move on the page moving the arrow keys. Let's say he presses the arrow-down key. I add one the the y1 coordinate and redraw the page.
This works very well with a PDF file that contains text (say, a PDF file generated with LibreOffice).
But, when the PDF contains images only (scanned A4 pages) I can't hover the rectangle around. It only moves within a small rectangle in the left corner of the image. It looks like I have two page, a big one and a small one. They are synchronized in the lower left corner.
![Rectangle](https://user-images.githubusercontent.com/69583002/135882928-a7fb049b-ffc5-4d27-8682-f158466bdd3d.png)
In can move the rectangle inside the smaller "page" and it corresponds to the positions in the bigger "page". But, when I use a regular PDF (text file), the two pages are synchronized.

Any hints, please?

**Reply by JorjMcKie** [reply]
I am pretty sure that this is caused by non-encapsulated "geometry changes" on the page.
Try `page.clean_contetns()` right after loading the page ...

**Reply by JorjMcKie** [reply]
Which means right after statement `pagina = doc[pno]`

**Reply by JorjMcKie** [reply]
of course: ``page.clean_contents()``

**Reply by FernandoJCabral** [reply]
> of course: `page.clean_contents()`

Sure!

Right on the spot. Thank you. Now it works perfectly well. You've solved in minutes a problem I had broke my head against for hours.
